packages/onboarding/src/getAssertData.ts

```diff
@@ -0,0 +1,519 @@
+  buff: Buffer,
+  {
+    maker,
+    connection,
+    owner,
+    rpcEndpoint,
+  }: { maker: PublicKey; connection: Connection; owner: PublicKey; rpcEndpoint: string },
+) => {
+  const walletDC = await getAssociatedTokenAddress(new PublicKey(DC_MINT), owner)
+
+  const makerDC = await getAssociatedTokenAddress(new PublicKey(DC_MINT), maker)
+
+  const [makerAccount, makerDcAccount, ownerAccount, ownerDcAccount] = await fetchSimulatedTxn({
+    apiUrl: rpcEndpoint,
+    accountAddresses: [maker.toString(), makerDC.toString(), owner.toString(), walletDC.toString()],
+    txnBuff: buff,
+  })
+
+  const fees = await estimateFees({
+    connection,
+    owner: {
+      key: owner,
+      account: ownerAccount,
+      dcAccount: ownerDcAccount,
+    },
+    maker: { key: maker, account: makerAccount, dcAccount: makerDcAccount },
+    dcMint: new PublicKey(DC_MINT),
+  })
+  return fees
+}
+
+const getAssertData = async ({
+  elevation: propsElevation,
+  decimalGain,
+  gateway,
+  nextLocation,
+  maker: propsMaker,
+  owner,
+  hotspotTypes,
+  onboardingClient,
+  connection,
+  hemProgram,
+  dcProgram,
+  cluster,
+}: {
+  hemProgram: HemProgram
+  connection: Connection
+  gateway: string
+  owner: PublicKey
+  decimalGain?: number
+  elevation?: number
+  maker?: PublicKey
+  nextLocation: string
+  hotspotTypes: HotspotType[]
+  onboardingClient: OnboardingClient
+  dcProgram: DcProgram
+  cluster: Cluster
+}) => {
+  const rpcEndpoint = connection.rpcEndpoint
+  const oraclePrice = await getOraclePriceFromSolana({ connection, cluster })
+  let solanaTransactions: Buffer[] | undefined
+  const gain = Math.round((decimalGain || 0) * 10.0)
+  const elevation = propsElevation || 0
+
+  const solanaAddress = owner.toBase58()
+  const balances = await getBalances(owner, connection)
+
+  const location = new BN(nextLocation, 'hex').toString()
+
+  let foundMaker = propsMaker
+  if (!foundMaker) {
+    const onboardingRecord = await onboardingClient.getOnboardingRecord(gateway)
+
+    if (onboardingRecord.data?.maker.address) {
+      foundMaker = heliumAddressToSolPublicKey(onboardingRecord.data?.maker.address)
+    }
+  }
+  if (!foundMaker) {
+    throw new Error('Could not determine hotspot maker')
+  }
+
+  const maker = foundMaker
+
+  const solResponses = await Promise.all(
+    hotspotTypes.map((type) =>
+      onboardingClient.updateMetadata({
+        type,
+        solanaAddress,
+        hotspotAddress: gateway,
+        location,
+        elevation,
+        gain,
+      }),
+    ),
+  )
+
+  solanaTransactions = solResponses
+    .flatMap((r) => r.data?.solanaTransactions || [])
+    .map((txn) => Buffer.from(txn))
+
+  let simulatedFees: (
+    | {
+        makerFees: {
+          lamports: number
+          dc: number
+        }
+        ownerFees: {
+          lamports: number
+          dc: number
+        }
+        isFree: boolean
+      }
+    | undefined
+  )[]
+  try {
+    simulatedFees = await Promise.all(
```

Comment:
I don't think this works in the case that the maker wallet has insufficient funds. The simulation will probably fail, and then it will put all fees on the owner. But the onboarding server does not have logic to change the payer, maker is still the payer not the user.

Reply:
@ChewingGlass We have this logic throwing an error in the simulation.
```ts
  if (response.data.result.value.err) {
    console.error(response.data.result.value.logs.join('\n'))
    throw new Error('Transaction would fail')
  }
  ```
  Is there any way to detect that it's the maker who doesn't have the funds here? Or would it be best to throw a specific error from the `update-metadata` onboarding server request?